File: tests/run_tests.py
"""Local evaluator for the Python starter.

Usage:
    python tests/run_tests.py --lang py
"""

from __future__ import annotations

import argparse
import importlib.util
import json
import sys
import time
from pathlib import Path


REPO_ROOT = Path(__file__).resolve().parent.parent


def load_agent(lang: str):
    if lang != "py":
        print(f"run_tests.py only supports --lang py (got {lang!r})", file=sys.stderr)
        sys.exit(2)
    agent_path = REPO_ROOT / "starter" / "py" / "agent.py"
    spec = importlib.util.spec_from_file_location("agent", agent_path)
    if spec is None or spec.loader is None:
        raise RuntimeError(f"cannot load {agent_path}")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    if not hasattr(module, "solve_ticket"):
        raise RuntimeError(f"solve_ticket not defined in {agent_path}")
    return module.solve_ticket


def score_resolution(res: dict, ticket: dict) -> float:
    if not isinstance(res, dict):
        return 0.0
    answer = res.get("answer", "")
    if not isinstance(answer, str):
        return 0.0
    signals = ticket.get("expected_signals") or []
    if not signals:
        return 1.0 if answer else 0.0
    hay = answer.lower()
    hits = sum(1 for s in signals if s.lower() in hay)
    return hits / len(signals)


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--lang", default="py")
    args = parser.parse_args()

    solve = load_agent(args.lang)
    tickets = json.loads((REPO_ROOT / "tests" / "tickets.json").read_text("utf-8"))

    total = 0.0
    for t in tickets:
        started = time.time()
        try:
            res = solve(t)
        except Exception as exc:
            print(f"[{t['id']}] threw: {exc}", file=sys.stderr)
            continue
        elapsed_ms = int((time.time() - started) * 1000)
        score = score_resolution(res, t)
        total += score
        conf = res.get("confidence") if isinstance(res, dict) else "?"
        print(f"[{t['id']}] score={score:.2f} time={elapsed_ms}ms conf={conf}")

    n = len(tickets)
    pct = (total / n) * 100 if n else 0
    print(f"\nTotal: {total:.2f} / {n} ({pct:.1f}%)")


if __name__ == "__main__":
    main()

File: tests/run_tests.js
// Shared local evaluator for the JS and TS starters.
// Usage: node tests/run_tests.js --lang js
//        node tests/run_tests.js --lang ts

import { readFile } from "node:fs/promises";
import { pathToFileURL } from "node:url";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));

function parseArgs(argv) {
  const out = { lang: "js" };
  for (let i = 0; i < argv.length; i++) {
    if (argv[i] === "--lang") out.lang = argv[++i];
  }
  return out;
}

function scoreResolution(res, ticket) {
  if (!res || typeof res.answer !== "string") return 0;
  const hay = res.answer.toLowerCase();
  const signals = ticket.expected_signals || [];
  if (signals.length === 0) return res.answer.length > 0 ? 1 : 0;
  const hits = signals.filter((s) => hay.includes(s.toLowerCase())).length;
  return hits / signals.length;
}

async function main() {
  const { lang } = parseArgs(process.argv.slice(2));
  const agentPath =
    lang === "ts"
      ? resolve(__dirname, "../starter/ts/agent.ts")
      : resolve(__dirname, "../starter/js/agent.js");

  const ticketsRaw = await readFile(
    resolve(__dirname, "tickets.json"),
    "utf8",
  );
  const tickets = JSON.parse(ticketsRaw);

  const mod = await import(pathToFileURL(agentPath).href);
  const solve = mod.solveTicket;
  if (typeof solve !== "function") {
    console.error(`solveTicket not exported from ${agentPath}`);
    process.exit(2);
  }

  let total = 0;
  for (const t of tickets) {
    const started = Date.now();
    let res;
    try {
      res = await solve(t);
    } catch (err) {
      console.error(`[${t.id}] threw:`, err.message);
      continue;
    }
    const ms = Date.now() - started;
    const score = scoreResolution(res, t);
    total += score;
    console.log(
      `[${t.id}] score=${score.toFixed(2)} time=${ms}ms conf=${res?.confidence ?? "?"}`,
    );
  }
  console.log(
    `\nTotal: ${total.toFixed(2)} / ${tickets.length} (${((total / tickets.length) * 100).toFixed(1)}%)`,
  );
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
